st2rbac_backend/resolvers.py

```diff
@@ -682,13 +684,96 @@ class KeyValuePermissionsResolver(PermissionsResolver):
+        log_context = {
+            "user_db": user_db,
+            "resource_db": resource_db,
+            "permission_type": permission_type,
+            "resolver": self.__class__.__name__,
+        }
+        self._log("Checking user resource permissions", extra=log_context)
+
+        # First check the system role permissions
+        has_system_role_permission = self._user_has_system_role_permission(
+            user_db=user_db, permission_type=permission_type
+        )
+
+        if resource_db.scope == FULL_SYSTEM_SCOPE and has_system_role_permission:
+            self._log("Found a matching grant via system role", extra=log_context)
+            return True
+
+        # Second check if this is for user scoped key-value pairs
+        key_prefix = user_db.name + ":"
+        if (resource_db.scope == "%s:%s" % (FULL_USER_SCOPE, user_db.name)) or (
+            resource_db.scope == FULL_USER_SCOPE and resource_db.name.startswith(key_prefix)
+        ):
+            self._log("Found default grant for the user scoped key value pair", extra=log_context)
+            return True
+
+        # Third check if this is a user scoped key-value pairs for another user
+        key_prefix = user_db.name + ":"
+        if resource_db.scope.startswith(
+            "%s:" % FULL_USER_SCOPE
+        ) and resource_db.scope != "%s:%s" % (FULL_USER_SCOPE, user_db.name):
+            self._log("User is trying to access another user's key value pairs", extra=log_context)
+            return False
+        if resource_db.scope == FULL_USER_SCOPE and not resource_db.name.startswith(key_prefix):
+            self._log("User is trying to access another user's key value pair", extra=log_context)
+            return False
+
+        # Check custom roles and permission grants
+        if permission_type in self.view_permission_types:
+            # Note: Some permissions such as "create", "modify", "delete" and "execute" also
+            # grant / imply "view" permission
+            permission_types = self.view_grant_permission_types[:] + [permission_type]
+            # Create role, permission grant for the permission types which doesn't exist
+            result = rbac_service.get_role_by_name(name=permission_type)
+            user_role = rbac_service.get_role_assignments_for_user(user_db, include_remote=True)
+            if len(user_role) > 0:
+               role = user_role[0].role
+            if result is None and (role == PermissionType.KEY_VALUE_PAIR_VIEW or role == PermissionType.KEY_VALUE_PAIR_LIST):
+                role_db = rbac_service.create_role(name=permission_type, description="Role of %s" % permission_type)
+                role_details = rbac_service.assign_role_to_user(role_db=role_db, user_db=user_db, source='assignments/%s.yaml' % user_db.name)
+                permission_grant = rbac_service.create_permission_grant_for_resource_db(role_db=role_db, resource_db=resource_db, permission_types=permission_types)        
```

Comment:
Why are we creating permission grant here? Which test case is this supporting. This looks very wrong.

Reply:
This was added to handle custom user roles with LIST/SET/DELETE for system scoped kvps.

Reply:
Custom user roles for specific system scoped KVPs should be defined explicitly in the YAML files. The resolvers should not be adding any new role assignment and permission grant records into the database automatically.

Reply:
Yes, that's correct and it works as mentioned.